src/jdk.naming.dns/share/classes/com/sun/jndi/dns/DNSDatagramChannelFactory.java

```diff
@@ -269,25 +260,29 @@ private DatagramSocket openRandom() {
                 if (family != null) {
                     DatagramChannel c = DatagramChannel.open(family);
                     try {
-                        DatagramSocket s = c.socket();
-                        s.bind(new InetSocketAddress(port));
-                        lastport = s.getLocalPort();
+                        c.bind(new InetSocketAddress(port));
+                        lastport = getLocalPort(c);
                         if (!recycled) history.add(port);
-                        return s;
+                        return c;
                     } catch (Throwable x) {
                         c.close();
                         throw x;
                     }
                 }
-                DatagramSocket s = new DatagramSocket(port);
-                lastport = s.getLocalPort();
+                var dc = DatagramChannel.open();
+                dc.bind(new InetSocketAddress(port));
+                lastport = getLocalPort(dc);
                 if (!recycled) history.add(port);
-                return s;
+                return dc;
             } catch (IOException x) {
                 // try again until maxtries == 0;
             }
```

Comment:
That can be simplified now:
```
                DatagramChannel dc = (family != null)
                                    ? DatagramChannel.open(family)
                                    : DatagramChannel.open();
                try {
                    dc.bind(new InetSocketAddress(port));
                    lastport = getLocalPort(dc);
                    if (!recycled) history.add(port);
                    return dc;
                }  catch (IOException x) {
                    dc.close();
                     // try again until maxtries == 0;
                }  
```
There's probably no need to retry if the open call fails, as the next call to `open` is likely to fail too. So if `open` throws we should probably let it propagate upwards - (and declare openRandom() to throw IOException).

Reply:
Thanks for suggestion - `openRandom` looks much simplier, no duplicated code and in-sync with `openDefault`. I agree that it is reasonable not to retry `open` call in failure case, and propagate `IOException`. Changed in fa7ef18713930aa19f1f45672ed3181acd2a0e63